test/built-ins/TypedArray/prototype/filter/speciesctor-destination-backed-by-immutable-buffer.js

```diff
@@ -0,0 +1,78 @@
+// Copyright (C) 2025 Richard Gibson. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+
+/*---
+esid: sec-%typedarray%.prototype.filter
+description: >
+  Throws a TypeError exception when the receiver constructs an instance backed
+  by an immutable buffer
+info: |
+  %TypedArray%.prototype.filter ( callback [ , thisArg ] )
+  ...
+  8. Repeat, while k < len,
+     a. Let Pk be ! ToString(𝔽(k)).
+     b. Let kValue be ! Get(O, Pk).
+     c. Let selected be ToBoolean(? Call(callback, thisArg, « kValue, 𝔽(k), O »)).
+     d. If selected is true, then
+        i. Append kValue to kept.
+        ii. Set captured to captured + 1.
+     e. Set k to k + 1.
+  9. Let A be ? TypedArraySpeciesCreate(O, « 𝔽(captured) », ~write~).
+  10. Assert: IsImmutableBuffer(A.[[ViewedArrayBuffer]]) is false.
+
+  TypedArraySpeciesCreate ( exemplar, argumentList [ , accessMode ] )
+  1. If accessMode is not present, set accessMode to ~read~.
+  2. Let defaultConstructor be the intrinsic object associated with the constructor name exemplar.[[TypedArrayName]] in Table 73.
+  3. Let constructor be ? SpeciesConstructor(exemplar, defaultConstructor).
+  4. Let result be ? TypedArrayCreateFromConstructor(constructor, argumentList, accessMode).
+
+  TypedArrayCreateFromConstructor ( constructor, argumentList [ , accessMode ] )
+  1. If accessMode is not present, set accessMode to ~read~.
+  2. Let newTypedArray be ? Construct(constructor, argumentList).
+  3. Let taRecord be ? ValidateTypedArray(newTypedArray, seq-cst, accessMode).
+
+  ValidateTypedArray ( O, order [ , accessMode ] )
+  1. If accessMode is not present, set accessMode to read.
+  2. Perform ? RequireInternalSlot(O, [[TypedArrayName]]).
+  3. Assert: O has a [[ViewedArrayBuffer]] internal slot.
+  4. If accessMode is ~write~ and IsImmutableBuffer(O.[[ViewedArrayBuffer]]) is true, throw a TypeError exception.
+features: [TypedArray, immutable-arraybuffer]
+includes: [testTypedArray.js, compareArray.js]
+---*/
+
+testWithAllTypedArrayConstructors((TA, makeCtorArg) => {
+  var calls = [];
+
+  var ta = new TA(makeCtorArg(["1", "2"]));
+  var constructor = {};
+  Object.defineProperty(ta, "constructor", {
+    get: function() {
+      calls.push("get ta.constructor");
+      return constructor;
+    }
+  });
+  Object.defineProperty(constructor, Symbol.species, {
+    get: function() {
+      calls.push("get ta.constructor[Symbol.species]");
+      var iab = (new TA(["3", "4"])).buffer.transferToImmutable();
+      var result = new TA(iab);
+      calls.push("construct result");
+      return result;
```

Comment:
I wondered why this test passes even without implementing the relevant changes in `%TypedArray%.prototype.filter`. The root of the problem is here. This `Symbol.species` getter doesn't return constructor (it returns a typed array). Not returning a constructor results in the expected `TypeError` (before we have any chance to get to the `IsImmutableBuffer()` that's supposed to be checked by this test).

In other words, the getter should not return `result` but a function that returns `result`.

The same problem is in the `slice` and `map` versions of this test.